### PR TITLE
Convert from Electron to Wails

### LINKED ISSUE #36: The installer doesn't execute
After downloading the new installer for windows I tried to run it but it doesn't work my cursor changes to the loading mode ig? but nothing work I tried changing the folder also tried disabling the anti-virus didn't work no luck with the compatibility mode either

### LINKED ISSUE #157: [EPIC] Snap and Flatpak compatibility
The BetterDiscord installer doesn't support Flatpak or Snap. The first Linux distros start seemlessly integrating these types of packages in their software store and as their main method of installation. 

Well, to be fair saying that the installer doesn't support snap isn't quite right either. Rather it explicitly blocks installing to snap even though it would otherwise be supported.
https://github.com/BetterDiscord/Installer/blob/1bf733dbb989c00ecf17a755ef931fa961195812/src/renderer/actions/paths.js#L84-L87
Flatpak on the other hand actually needs some more work because of it's sandboxing by the looks of it. More specifically it can't read the `betterdiscord.asar` from user's config directory, so it needs to be installed in the flatpak's config directory.

The snap installation has a related problem. because it's `betterdiscord.asar` is in the user config directory while the detected config directory is in the snap package's directory the autoupdate functionality of BD is expected to fail.

From a chat with @Tropix126 I could gather that it wasn't always compatible but changing it to injection method to inject the `require` statement in the `discord_desktop_core/index.js` module seems to have made it otherwise compatible. In fact [`betterdiscordctl`](https://github.com/bb010g/betterdiscordctl)'s installation method is identical to the installer. The only relevant change for BD installation is changing the config directory detection depending on the detected. 

<details> 
  <summary>betterdiscordctl's installation steps for comparison (warning: identical to the installer)</summary>
  <ol>
    <li>Detect config directory</li>
    <li>Detect Discord folder in the config directory</li>
    <li>Detect version folder in the Discord config directory</li>
    <li>Detect the core module folder in the version directory</li>
    <li>download betterdiscord.asar from github releases (it looks like it may also be caching the asar though)</li>
    <li> create a betterdiscord and betterdiscord/data folder if they don't exist</li>
    <li>copy the asar in there</li>
    <li>inject require statement in index.js of the core module (only difference I can see is a newline more after the injection using betterdiscordctl)</li>
  </ol>
</details>

- [ ] Techincal Requirements
  - [ ] Snap support
    - [ ] version folder detection (trivial with #150 + @ObserverOfTime's snippet from #26)
    - [ ] remove explicit block from the path verification
    - [ ] install to betterdiscord.asar to the package's home directory
      - [ ] check if package update removes the BD config folderwhen doing so
  - [ ] Flatpak support
    - [ ] version folder detection (same as snap)
    - [ ] install to betterdiscord.asar to the package's home directory
      - [ ] check if package update removes the BD config folderwhen doing so
  - [ ] BetterDiscord compatibility?
    - [ ] BetterDiscord seems to function fine on snap and flatpak but maybe there are details that don't function correctly because they require accessing files outside of the BD config directory
- [ ] Design/UX considerations
  - [ ] should the list be exanded to include snap/flatpak?
  - [ ] should snap/flatpak be fallback if for automatic detection if the native installation is not found?
  - [ ] should snap/flatpak be in separate platform selection dialogs?
    - [ ] if so how should those be accessed?
      - [ ] **keep in mind lots of people won't know what type of installation they have because the software stores of their distro's make it seemless.** they probably won't even know what the hell snap/flatpak is. but would still be nice to choose which variant to install to
  - [ ] technical: adjustments to internal data structure to keep it clean
    - [ ] internal platform name + folderName + display name + image url separate, so you don't need to do `.toLowercase().replace(" ","")` everywhere
      - [ ] also don't hardcode the existing platforms everywhere but base them of a common object where all are defined, this would also allow adding a `Discord - Snap` and `Discord - Flatpak` platform as needed,

### LINKED ISSUE #202: Add options to repair and uninstall
**Is your feature request related to a problem? Please describe.**
kinda, like if a plugin/theme is having issues the installer can have a option in repair that would disable all plugins and themes

**Describe the solution you'd like**
Add options to repair and uninstall
Like repair can have the options of `disable addons and reinstall bd` (addons being themes and plugins) and `reinstall bd` and uninstall can have a way to fully remove bd and just remove it off of discord

**Describe alternatives you've considered**
Basically something like those

**Additional context**
no

### LINKED ISSUE #226: [Bug] Linux: Stuck installing "Downloading asar file"
**Describe the Bug**
When trying to install using the `.AppImage` it gets stuck on the installation and never finishes or does anything.

**To Reproduce**
Simply trying to install.

**Expected Behavior**
To fully install

**Screenshots**
Issue I'm seeing while installing:

Starting Install...

Creating required directories...
✅ Directory created: /home/CodeOnBlocks/.config/BetterDiscord
✅ Directory created: /home/CodeOnBlocks/.config/BetterDiscord/data
✅ Directory created: /home/CodeOnBlocks/.config/BetterDiscord/themes
✅ Directory created: /home/CodeOnBlocks/.config/BetterDiscord/plugins
✅ Directories created

Downloading asar file

**Discord Version**
Stable

I'm on openSUSE Tumbleweed

### LINKED ISSUE #359: [Bug] betterdiscord installer keeps crashing
### Before Reporting

- [X] I found no existing issues matching my bug
- [X] My issue is not caused by a plugin or theme

### Describe the Bug

everytime i have tried to open the betterdiscord installer it instantly crashes after a few seconds. it suddenly happened a few months ago and it wasn't really a problem until the new update where i just can't update now at all. i've tried running the installer on several virtual machines and it works just fine but on my main computer it just doesn't work at all. any way to fix this?

### Expected Behavior

betterdiscord installer works and installs to 1.9.4

### Steps To Reproduce

1. run betterdiscord installer
2. try to update via the updater on the betterdiscord menu on discord

### Operating System

Windows

### Discord Version

Stable

### BetterDiscord Version

1.9.3

### Anything else?

_No response_

### LINKED ISSUE #362: [Bug] BD Installer hangs after "Downloading asar file"
### Before Reporting

- [X] I found no existing issues matching my bug
- [X] My issue is not caused by a plugin or theme

### Describe the Bug

Using the BetterDiscord app image on Fedora 39 it seems that every attempt at installing BetterDiscord leads to an installer hang. Running the app image through the terminal does not give any notable output, it just simply hangs. Sometimes it complains that it failed to download the package from the official website giving a 502 error, sometimes it just hangs while downloading the asar file.

Usual output before hanging
```
Downloading asar file
❌ Failed to download package from the official website
❌ Status code did not indicate success: 502
Falling back to GitHub...
```

### Expected Behavior

Better discord installer works

### Steps To Reproduce

1. Download app image
2. Run app image
3. Go through install process
4. See hang on install

### Operating System

Linux

### Discord Version

Stable

### BetterDiscord Version

v1.3.0

### Anything else?

Installer window is still responsive, it just stops doing anything.

### LINKED ISSUE #365: Autodetection of Flatpak Discord
**Is your feature request related to a problem? Please describe.**
I'd like to be able to install betterdiscord without having to manually select the long flatpak directory path.

**Describe the solution you'd like**
Could you auto detect the flatpak version of Discord, just like the normal version? It installs fine when manually selecting the directory `~/.var/app/com.discordapp.Discord/config/discord/0.0.x`

**Describe alternatives you've considered**
Typing out the entire thing every time (have been doing this)

**Additional context**
image

### LINKED ISSUE #366: The installer gets stuck trying to terminate the discord service
Hello,

To play this I double-click on the installer, accept the terms, click next, select the first one, click next, click the first Discord option, and finally click Install.
The installation starts, but does not finish, because it tries to end Discord, but nothing. I have tried restarting, reinstalling Discord from scratch, uninstalling the antivirus, deleting registry files with the name Discord, deleting all the folders on the PC that contain the name Discord but nothing.

**Discord Version**
Stable

**Operating System**
Windows 10 Pro

Win10

### LINKED ISSUE #367: Improve Keyboard Navigation
Expected behavior:
- TAB cycles through interactive UI elements, focusing them (already the case)
- SPACE activates/toggles the focused UI element (already the case)
- ENTER *submits* the current form/view/page (currently behaves identical to SPACE press)

### LINKED ISSUE #375: Title bar being removed/updated on pressing back button
**Describe the bug**
The title for the installer will disappear on the taskbar and when hovering over the application.
image
image

**To Reproduce**
1. Open up the installer
2. Accept license agreement and go to next page
3. Press 'back' button
4. The title will have disappeared.

**Solution**
*I didn't think this was worth making my own PR so I'll just leave the solution here*
In the createMainWindow function in index.js you can add this code to prevent the title from being updated whatsoever after window creation. This mitigates the issue.
```
// Prevents the page title from being updated
window.on('page-title-updated', (e) => {
    e.preventDefault();
});
```

### LINKED ISSUE #377: [Bug] Discord Canary not found in BD Installer on Windows after reinstall
### Before Reporting

- [x] I found no existing issues matching my bug
- [X] My issue is not caused by a plugin or theme

### Describe the Bug

I recently performed a full uninstall of Discord Canary with Revo Uninstaller to remove all temp files and registry keys.

I then ran the Discord Canary installer, `DiscordCanarySetup.exe`. 

Discord Canary runs and was installed in the expected location ( `%LocalAppData%\DiscordCanary` ).

However when I run BetterDiscord Installer v1.3.0 it does not show Discord Canary as found, even when I manually click 'Browse' to select the folder. 

I tried clicking 'Browse' and navigating to both `%LocalAppData%\DiscordCanary` and `%LocalAppData%\DiscordCanary\app-1.0.92`.

image

Before the un/re-install the BetterDiscord installer would show Discord Canary as expected.

### Expected Behavior

If Discord Canary is installed in the default location ( `%LocalAppData%\DiscordCanary` ) then the BetterDiscord Installer should see it.

### Steps To Reproduce

I did a full Discord Canary Uninstall with Revo Uninstaller, deleting all relevant files and registry keys, and then reinstalled it using `DiscordCanarySetup.exe` and allowed it to update to the latest version.

### Operating System

Windows

### Discord Version

Canary

### BetterDiscord Version

v1.3.0

### Anything else?

_No response_

### LINKED ISSUE #379: [Bug] Linux installer doesn't run out of the box on Ubuntu 24.04
### Before Reporting

- [X] I found no existing issues matching my bug
- [X] My issue is not caused by a plugin or theme

### Describe the Bug

The Linux AppImage installer fails to run on a base `amd64` installation of Ubuntu Linux 24.04 LTS. Here is the error message:

```
$ ./BetterDiscord-Linux.AppImage 
dlopen(): error loading libfuse.so.2

AppImages require FUSE to run. 
You might still be able to extract the contents of this AppImage 
if you run it with the --appimage-extract option. 
See https://github.com/AppImage/AppImageKit/wiki/FUSE 
for more information
```

This is because by default Ubuntu 22.04 comes with `libfuse3.so.3` instead, from the `libfuse3-3` package.

For now, the missing shared library can still be installed using `sudo apt install libfuse2t64`. It might be helpful to document this extra step.

Also, I haven't checked whether this is specific to Ubuntu, or it came from upstream Debian. It might also affect recent versions of other distros.

### Expected Behavior

AppImage binary should work out of the box.

### Steps To Reproduce

1. Download the installer from the official project website
2. Mark it as executable with `chmod +x`
3. Try to execute it.

### Operating System

Linux

### Discord Version

Stable

### BetterDiscord Version

N/A

### Anything else?

_No response_

### LINKED ISSUE #380: [Feature] Native Apple silicon build support
### Before Requesting

- [X] I found no existing issue matching my feature request

### Describe the feature you'd like!

An available native Apple silicon build.

### Anything else?

Currently, we still need to install Rosetta 2 in order for BetterDiscord to work.

### LINKED ISSUE #385: [Workaround] Workaround script to install BetterDiscord for Discord installations with flatpak
I've been waiting for the install update for flatpak but I didn't get a satisfactory outcome. So I did a quick workaround and I decided to share it with you.

Requirements:

1. bash (usually installed)
2. jq [(may need installation)](https://jqlang.org/download/)
3. curl (usually installed)
4. sed (usually installed)

script available here: https://gist.github.com/caravelaPortuguesa/c7d41674ea4709f50b5d844f266a56a5
And also here:
```
#!/bin/bash

shopt -s nullglob
discord_flatpak_base="$HOME/.var/app/com.discordapp.Discord/config"

set -x
potentials=(
	"$discord_flatpak_base/discord/"[0-2]*.[0-9]*.[0-9]*"/modules/discord_desktop_core"*/index.js
)

for potential in "$potentials"; do

	if ! grep -P 'required\([a-z/]+BetterDiscord/data/betterdiscord.asar' "$potential" ; then
		dl_url="$(curl -s https://api.github.com/repos/BetterDiscord/BetterDiscord/releases/latest | jq -r '.assets[] | select(.name | contains(".asar")) | .browser_download_url')"
		curl -L "$dl_url" -o "$discord_flatpak_base"/BetterDiscord/data/betterdiscord.asar

		sed -i '1s@^@require("../../../../BetterDiscord/data/betterdiscord.asar");\n@' "$potential"
	fi
done
```

### LINKED ISSUE #408: [Bug] BD does not detect Discord on macOS
### Before Reporting

- [x] I found no existing issues matching my bug
- [x] My issue is not caused by a plugin or theme

### Describe the Bug

During installation, BD is unable to detect Discord on macOS since the latest version.
`canary 535274 (c7e8a46) Host 0.0.1089 arm64 (81192) Build Override: N/A macOS 26 (25.5.0)`

### Expected Behavior

Discord must be detected by the installer.

### Steps To Reproduce

1. Launch installer
2. Install BD

### Operating System

MacOS

### Discord Version

Canary

### BetterDiscord Version

1.13.11

### Anything else?

_Upon launch, it creates two version folders ("X.X.XXXX" and "app-X.X.XXXX"). I think the problem stems from this. As soon as I reinstall the app, the "app-X.X.XXXX" folder disappears, but it reappears on the second launch. When I perform a factory reset without the "app-X.X.XXXX" folder, the installer correctly detects Discord and it works, but when I relaunch the app, the "app-X.X.XXXX" folder reappears and BD uninstalls._

### LINKED ISSUE #409: [Bug] Discord no longer autodetected when installing
### Before Reporting

- [x] I found no existing issues matching my bug
- [x] My issue is not caused by a plugin or theme

### Describe the Bug

When updating Discord to 1.0.136, BetterDiscord cannot see the Discord client to reinstall. 

### Expected Behavior

BetterDiscord sees the Discord client automatically; or it can be selected by following the steps detailed in the Troubleshooting documentation

### Steps To Reproduce

Update Discord to 1.0.136
Run BetterDiscord installer
When choosing which Discord version, no versions are autodetected, and cannot by selected by browsing, even within Discord's files. 

### Operating System

Linux

### Discord Version

Stable

### BetterDiscord Version

v1.13.13 (assumed, cannot load it)

### Anything else?

I have tried reinstalling both Discord and the BetterDiscord Installer. I have also gone through the troubleshooting documentation and tried to complete it to the best of my ability (am on Linux and solutions are describe for Windows)

### LINKED ISSUE #410: Cant Download
So ive removed my vencord to use bd. i pressed remove vencord and its gone from discord as well. when I downloaded the bd installer and ran it, it kept on getting to "attempting to kill discord" message and did nothing. and no discord is open i checked task manager. ive even deleted my discord folder in my local and also reinstalled discord.

### PR DESCRIPTION
**Pull request by zerebos**
# Overview
This is part of the effort to convert from an Electron-based application to a Wails-based application. It would reduce the number of issues on Windows with the weird electron-forge zip-wrapping, and massively reduce both the filesize and runtime overhead. This can lay the groundwork to share common code between this and our CLI. As a bonus, if we ever decide to go forward with the "bd manager" app idea, this would pave the way there as well.

## Important things to do before merging:
- [x] Update README
- [x] Update CONTRIBUTING
- [x] Add a sanity check CI for main and PRs
- [x] Adjust release pipeleine to properly build and release
- [x] Add update check
- [x] Include the linux love and care changes by working with @samfundev
  - [x] Expose flatpak to global BetterDiscord folder and make BD use `HOST_XDG_CONFIG_HOME`
  - [x] Install locally for snap installs
- [x] General code cleanup
  - [x] Separate the non-common action code into postfixed files like `actions_windows.go`
- [x] Bug testing with beta users
  - [x] Testing general cases
  - [x] Testing edge cases
- [x] Unit tests
  - [x] backend tests in golang
  - [x] frontend tests in typescript
- [x] Update dependencies like svelte to v5 (it has been a long time since I worked on this)
- [x] Add build script to build for all of our intended os/arch combos
- [x] Make `Enter` submit the page
- [x] Get confirmation on our approach to snap and flatpak

## Can be done later after merging:
- [ ] Update to wails v3
- [x] Add more options to repair (#202)
- [ ] Code signing on windows and notarization on mac (#390, #126)
- [ ] Translations (#16)
  - [ ] BetterDiscord gets setup on Crowdin starting on client
  - [ ] Determine whether to do joint i18n across process or pass the buck to one side
  - [ ] Move strings to Crowdin

## These have been decided against for this project:
- ~~Add automatic reinjection~~
- ~~Check for addon updates~~
- ~~Add system tray~~
- ~~Add start at login~~
- ~~Built in CLI (#357)~~ BD now has a comprehensive but separate cli

## This PR affect the following issues:
- Hopefully fixes #410 by doing a new method
- Fixes #409 by supporting new updater
- Fixes #408 with new updater detection
- Fixes #385 by properly supporting flatpak
- Fixes #377 by doing proper folder checks with fallback
- Fixes #375 
- Fixes #379 by no longer using that library
- Fixes #380 because cross compilation with Go is dead simple
- Fixes #367 by adjusting keyboard nav
- Hopefully fixes #366 by using native methods
- Fixes #365 by properly supporting flatpak
- Hopefully fixes #362 by using native methods
- Fixes #359 and fixes #36 with the new architecture
- Hopefully fixes #226 by using native methods
- Closes #157 by properly supporting both
- Implements and closes #202 

## Preview of the new configs
https://github.com/user-attachments/assets/84ce2664-a359-47ad-82e0-b54b3514062b

This has been updated to include checkboxes for accessibility:

image

### CODE REVIEW

File: betterdiscord/download.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+func (i *BDInstall) download() error {
+	if i.hasDownloaded {
+		log.Printf("✅ Already downloaded to %s", i.asar)
+		return nil
+	}
+
+	resp, err := utils.DownloadFile("https://betterdiscord.app/Download/betterdiscord.asar", i.asar)
+	if err == nil {
+		version := resp.Header.Get("x-bd-version")
+		log.Printf("✅ Downloaded BetterDiscord version %s from the official website", version)
```
[the remark this answers is not in the document]
Comment (zerebos) [reply betterdiscord/download.go]: Isn't this set on line 59 of this file?

File: betterdiscord/setup.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+	}
+
+	contents, err := os.ReadFile(path)
+	if err != nil {
+		log.Printf("❌ Unable to read %s: %s\n", label, path)
+		log.Printf("   %s\n", err.Error())
+		return err
+	}
+
+	backupPath := filepath.Join(backupDir, filepath.Base(path))
```
Comment (doggybootsy) [inline betterdiscord/setup.go:145]: Should something like the current time be used in the filename too? Like for example, someone does clear custom css twice (by just repetition/accidentally) it will just nuke their css, when they probably didn't want that to happen.

Comment (zerebos) [reply betterdiscord/setup.go:145]: I could push the name generation down to here, but if you look at line 47, each repair gets its own subfolder with a timestamp postfixed name like `backup/repair-20260221-235932` so if they do another repair immediately after the backups will be in a new subfolder

Comment (doggybootsy) [reply betterdiscord/setup.go:145]: oh no need to push it down here. I didn't notice it did that up there

Comment (zerebos) [reply betterdiscord/setup.go:145]: image

**Review by doggybootsy — APPROVED** [review]
Forgot to approve this. LGTM

**Comment by zerebos** [comment]
Tested working on Windows, macOS, and Linux (Fedora). Was able to test both new and old updater on macOS and Linux. On Linux I tested native, snap, and flatpak installs.

**Review by doggybootsy — APPROVED** [review]
From my understanding of the code, it looks good to me
